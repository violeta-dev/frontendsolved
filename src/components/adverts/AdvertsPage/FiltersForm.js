import React from 'react';
import T from 'prop-types';
import { connect } from 'react-redux';
import { Button, Input, Slider, Radio, Row, Col } from 'antd';

import TagsSelect from '../TagsSelect';
import FormField from '../../shared/FormField';
import { saleOptions, MIN_PRICE, MAX_PRICE } from '../../../definitions';
import styles from './FiltersForm.module.css';
import { loadFilteredAdverts } from '../../../store/actions';
import { getFilters } from '../../../store/selectors';

export class FiltersForm extends React.Component {
  state = {
    ...this.props.initialFilters,
  };

  handleNameChange = ev => this.setState({ name: ev.target.value });
  handlePriceChange = price => {
    const [min, max] = price;
    if (min === MIN_PRICE && max === MAX_PRICE) {
      return this.setState({ price: [] });
    }
    this.setState({ price });
  };
  handleSaleChange = ev => this.setState({ sale: ev.target.value });
  handleTagsChange = tags => this.setState({ tags });

  handleSubmit = ev => {
    const { onSubmit } = this.props;
    ev.preventDefault();
    onSubmit(this.state);
  };

  render() {
    const { name, price, tags, sale } = this.state;
    const priceValue = price.length === 0 ? [MIN_PRICE, MAX_PRICE] : price;

    return (
      <form onSubmit={this.handleSubmit}>
        <Row className={styles.form}>
          <Col span={11}>
            <FormField label="By name">
              <Input
                placeholder="Name"
                onChange={this.handleNameChange}
                value={name}
              />
            </FormField>
            <FormField
              label={
                <>
                  By price
                  <strong style={{ margin: '0 5px' }}>
                    {priceValue.join(' - ')}
                  </strong>
                </>
              }
            >
              <Slider
                range
                defaultValue={priceValue}
                min={MIN_PRICE}
                max={MAX_PRICE}
                onChange={this.handlePriceChange}
              />
            </FormField>
          </Col>
          <Col span={11} offset={2}>
            <FormField label="By tags">
              <TagsSelect onChange={this.handleTagsChange} value={tags} />
            </FormField>
            <FormField label="By type">
              <Radio.Group
                options={Object.values(saleOptions)}
                onChange={this.handleSaleChange}
                value={sale}
              />
            </FormField>
          </Col>
          <Col span={24}>
            <Button
              className={styles.button}
              type="primary"
              htmlType="submit"
              block
            >
              Search
            </Button>
          </Col>
        </Row>
      </form>
    );
  }
}

FiltersForm.propTypes = {
  initialFilters: T.shape({
    name: T.string.isRequired,
    sale: T.oneOf(Object.keys(saleOptions)).isRequired,
    price: T.arrayOf(T.number).isRequired,
    tags: T.arrayOf(T.string).isRequired,
  }).isRequired,
  onSubmit: T.func.isRequired,
};

const mapStateToProps = state => ({ initialFilters: getFilters(state) });

const mapDispatchToProps = dispatch => ({
  onSubmit: filters => dispatch(loadFilteredAdverts(filters)),
});

export default connect(mapStateToProps, mapDispatchToProps)(FiltersForm);
