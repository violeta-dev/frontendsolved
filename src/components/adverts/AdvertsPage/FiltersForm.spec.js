import React from 'react';
import { shallow } from 'enzyme';

import { FiltersForm } from './FiltersForm';

describe('FiltersForm', () => {
  const props = {
    onSubmit: jest.fn(),
    initialFilters: { name: 'Car', sale: 'all', price: [0, 1000], tags: [] },
  };
  const render = () => shallow(<FiltersForm {...props} />);
  test('snapshot', () => {
    const wrapper = render();
    expect(wrapper).toMatchSnapshot();
  });

  test('should submit form with filters', () => {
    const wrapper = render();

    expect(
      wrapper.find('FormField').at(0).children().first().props().value,
    ).toBe(props.initialFilters.name);
    expect(
      wrapper.find('FormField').at(1).children().first().props().defaultValue,
    ).toBe(props.initialFilters.price);
    expect(
      wrapper.find('FormField').at(2).children().first().props().value,
    ).toBe(props.initialFilters.tags);
    expect(
      wrapper.find('FormField').at(3).children().first().props().value,
    ).toBe(props.initialFilters.sale);

    wrapper.find('form').simulate('submit', { preventDefault: () => {} });
    expect(props.onSubmit).toHaveBeenCalledWith(props.initialFilters);
  });
});
