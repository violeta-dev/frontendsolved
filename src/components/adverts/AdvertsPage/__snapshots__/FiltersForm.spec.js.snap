// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FiltersForm snapshot 1`] = `
<form
  onSubmit={[Function]}
>
  <Row
    className="form"
  >
    <Col
      span={11}
    >
      <FormField
        label="By name"
      >
        <Input
          onChange={[Function]}
          placeholder="Name"
          type="text"
          value="Car"
        />
      </FormField>
      <FormField
        label={
          <React.Fragment>
            By price
            <strong
              style={
                Object {
                  "margin": "0 5px",
                }
              }
            >
              0 - 1000
            </strong>
          </React.Fragment>
        }
      >
        <Slider
          defaultValue={
            Array [
              0,
              1000,
            ]
          }
          max={10000}
          min={0}
          onChange={[Function]}
          range={true}
          tipFormatter={[Function]}
        />
      </FormField>
    </Col>
    <Col
      offset={2}
      span={11}
    >
      <FormField
        label="By tags"
      >
        <Connect(TagsSelect)
          onChange={[Function]}
          value={Array []}
        />
      </FormField>
      <FormField
        label="By type"
      >
        <Memo([object Object])
          onChange={[Function]}
          options={
            Array [
              Object {
                "label": "Sell",
                "value": "sell",
              },
              Object {
                "label": "Buy",
                "value": "buy",
              },
              Object {
                "label": "All",
                "value": "all",
              },
            ]
          }
          value="all"
        />
      </FormField>
    </Col>
    <Col
      span={24}
    >
      <Button
        block={true}
        className="button"
        ghost={false}
        htmlType="submit"
        loading={false}
        type="primary"
      >
        Search
      </Button>
    </Col>
  </Row>
</form>
`;
